pytest_mh/utils/tc.py

```diff
@@ -0,0 +1,74 @@
+from __future__ import annotations
+
+from .. import MultihostHost, MultihostUtility
+from ..ssh import SSHLog
+
+__all__ = ["LinuxTrafficControl"]
+
+
+class LinuxTrafficControl(MultihostUtility):
+    """
+    Perform traffic control operations on remote host.
+
+    All changes are automatically reverted when a test is finished.
+    """
+
+    def __init__(self, host: MultihostHost) -> None:
+        """
+        :param host: Remote host instance.
+        :type host: MultihostHost
+        """
+        super().__init__(host)
+        self.__interfaces: set[str] = set()
+        self.__restore: set[str] = set()
+
+    def setup_when_used(self):
+        """
+        Setup traffic control configuration.
+
+        :meta private:
+        """
+        super().setup_when_used()
+
+        # Let's find out the availabel interfaces
+        result = self.host.ssh.run("ip -o address", log_level=SSHLog.Error)
+        split_result = result.stdout.splitlines()
+
+        for line in split_result:
+            interface = line.split(" ")[1]
+            if interface != "lo":
+                self.__interfaces.add(interface)
+
+    def teardown_when_used(self):
+        """
+        Revert all traffic control changes.
+
+        :meta private:
+        """
+        tear: str = ""
+        for iter in self.__restore:
+            tear += iter + "\n"
+        self.host.ssh.run(tear, log_level=SSHLog.Error)
+        super().teardown_when_used()
+
+    def add_delay(self, hostname: str, time: int):
+        """
+        Add delay to network connection.
+
+        :param hostname: Target hostname.
+        :type hostname: str
+        :param time: Delay in milliseconds.
+        :type time: int
+        """
+        ip = self.host.ssh.run(f"dig +short {hostname}", log_level=SSHLog.Error)
+
+        for interface in self.__interfaces:
+            self.host.ssh.run(
+                f"""
+                set -e
+                tc qdisc add dev {interface} root handle 1: prio
+                tc qdisc add dev {interface} parent 1:3 handle 30: netem delay {time}
+                tc filter add dev {interface} protocol ip parent 1:0 priority 3 u32 match ip dst {ip.stdout} flowid 1:3
```

Comment:
This works. But it is not flexible, because it allows you to only add single delay. Also this delay does not apply only to the particular IP address but also any other packet which TOS sends it to 1:3 band of the priority queue.

So, we need to do some changes.

* We need to be handle-id aware, we can't just use hardcoded values.
* We must be able to create the prio root qdisc separately, because we need to set the number of bands to support. The default value is 3, but that can handle only two distinct delay (default, custom1, custom2).
* All non-filtered packets must be set to band 1, other packets must be send to specified delays.
* We still want to provide some simple API.

This is starting to be a challenge :-)

1. Adding a prio qdisc with custom number of bands (10 in this example), all packets are send to band 0, handle id is `1:`
```
tc qdisc add dev eth0 root handle 1: prio bands 10 priomap 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
```

2. Adding a netem child to set delay, this child is added to band `1` (`1:2` means qdisc `1` band `2-1=1`), this child has handleid `30`
```
tc qdisc add dev eth0 parent 1:2 handle 30: netem delay 150ms
```

3. Add filter that will send specific packet to band 1 `priority 2 ... flowid 1:2` (all other packets are in band 0 = without delay)
```
tc filter add dev eth0 protocol ip parent 1: priority 2 u32 match ip dst 172.16.100.10 flowid 1:2
```

Repeating steps 2 and 3 with can define different delays for different bands.

Reply:
This however starts to be over-complicated, we would need to add individual methods like `qdisc_add_prio`, `qdisc_add_netem`, `add_filter` and so on.

I think the most common use case is to set a delay to all outgoing connections, so it would be enough to stick to the current version (with some modification - you still need to set all unfiltered packets to one band) and say you can call it only once (define the limitations). Providing a replace/delete method if one needs to change the delay. WDYT?

Reply:
More generic API can always be created later, or we can add a method to just run tc command with user given arguments given the command complexity. We can just take care of the teardown and obtaining list of interfaces.

Reply:
Actually, the maximum amount of bands is 16 so we can hardcode this. Then we can support up to 15 delays.

* create prio queue (if not exist) with 16 bands, send everything to 0
* add a delay and assign a band to it, remember the band
* add a filter and assign it to band based on requested delay

Some delete API should also be handy to remove particular filter.

Reply:
I think I understood what you mean in the previous comments, but let me explain what I did.

`add_delay()` keeps the same interface as before, but currently I set a limitation to the number of connections that can be delayed at a time to 15. Band 0 is used for the default and non-delayed connections. Band 1 to 15 are used for delayed traffic, and each hostname that adds a delay uses a band. It is recommended to set the delays from minimum to maximum to avoid starvation. All this information is provided in the API comments.

Another option that came to my mind and that can increase the number of delays in parallel, is to divide the bands per delay range. Thus, a delay of 1-100ms could be assigned to band 1, a delay of 101-200ms band 2, etc. I don't mind implementing that as it would give us more flexibility and the user doesn't have to be aware of any limitation. In any case, do you think we'll be using more than 1 or 2 delays at a time?

Regarding the API to remove a particular filter, let's agree on the delay addition first. It already took me some time to figure this out and I think I'll have to change it following your feedback.

Finally, please next time try to avoid having several contradictory comments without a final statement of what is actually expected (or isn't). It was difficult to follow the discussion and I had to read it several times.